### ISSUE TITLE
JavaScript alert triggered, breaking build

### ISSUE THREAD

**Issue by Koronen** [thoughtbot/capybara-webkit#729]
Originally opened as thoughtbot/hound#403.

@Koronen: 
> After checking out the current `master` branch (thoughtbot/hound@b292540f2f5535a8fea4d72fce095de2f6b909c9) and running either `bundle exec rake`, or a more narrowed down `bundle exec rspec spec/features/repo_list_spec.rb`, I almost always get a JavaScript alert saying "Your repos failed to sync." (see screenshot), pausing execution of the tests. After I press OK, all remaining tests in that file (and in the case of `bundle exec rake` also some tests in other files) fail.

@jferris:
> I have the same issue, and I'm not sure how it's happening. We don't implement a GUI alert like that in capybara-webkit, so I'm not sure how it's managing to make one! Can you create an issue on capybara-webkit?

This issue is still present on my machine running tests for the current `master` branch of Hound (thoughtbot/hound@87d968266dcc185b07c036a6579cee5a964e2211 at the time of this writing). Given Joe's reply, this appears to be a bug in capybara-webkit. Any clues as to what could be causing this?

![hound_b292540f2f5535a8fea4d72fce095de2f6b909c9_test_failure](https://cloud.githubusercontent.com/assets/543455/4407505/d1016c94-44ca-11e4-82fe-92f80eac0574.png)

```
Failures:

  1) Repo list user deactivates private repo without subscription
     Failure/Error: sign_in_as(user)
     EOFError:
       end of file reached
     # ./spec/support/helpers/authentication_helper.rb:11:in `sign_in_as'
     # ./spec/features/repo_list_spec.rb:132:in `block (2 levels) in <top (required)>'

  2) Repo list user syncs repos
     Failure/Error: sign_in_as(user)
     Errno::EPIPE:
       Broken pipe
     # ./spec/support/helpers/authentication_helper.rb:11:in `sign_in_as'
     # ./spec/features/repo_list_spec.rb:35:in `block (2 levels) in <top (required)>'

  3) Repo list user views list
     Failure/Error: sign_in_as(user)
     Errno::EPIPE:
       Broken pipe
     # ./spec/support/helpers/authentication_helper.rb:11:in `sign_in_as'
     # ./spec/features/repo_list_spec.rb:9:in `block (2 levels) in <top (required)>'

  4) Repo list user filters list
     Failure/Error: sign_in_as(user)
     Errno::EPIPE:
       Broken pipe
     # ./spec/support/helpers/authentication_helper.rb:11:in `sign_in_as'
     # ./spec/features/repo_list_spec.rb:21:in `block (2 levels) in <top (required)>'

  5) Repo list user with admin access activates organization repo
     Failure/Error: sign_in_as(user)
     Errno::EPIPE:
       Broken pipe
     # ./spec/support/helpers/authentication_helper.rb:11:in `sign_in_as'
     # ./spec/features/repo_list_spec.rb:93:in `block (2 levels) in <top (required)>'

  6) Repo list user activates repo
     Failure/Error: sign_in_as(user)
     Errno::EPIPE:
       Broken pipe
     # ./spec/support/helpers/authentication_helper.rb:11:in `sign_in_as'
     # ./spec/features/repo_list_spec.rb:67:in `block (2 levels) in <top (required)>'
```

```
koronen@paridae$ ruby --version
ruby 2.2.0p0 (2014-12-25 revision 49005) [x86_64-linux]
koronen@paridae$ qmake --version
QMake version 2.01a
Using Qt version 4.8.6 in /usr/lib/x86_64-linux-gnu
koronen@paridae$ uname -a
Linux paridae 3.13.0-45-generic #74-Ubuntu SMP Tue Jan 13 19:36:28 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by stringsn88keys** [comment]
I'm wondering if this is related to the random `EOFError` that I get when tagging scenarios in cucumber with `@javascript`. With version 1.3.0 of this gem, I get the error. With version 1.1.0 of the gem, I don't get the error.

**Comment by jferris** [comment]
@stringsn88keys any chance you have a minimal test case that shows the problem in 1.3.0 and not 1.1.0? Also, do you see the problem in 1.2.0? If we're doing a source search for the regression, it would greatly help to narrow the range and have something reproducible.

**Comment by jferris** [comment]
This seems to happen when:

* There are two examples run in a row
* The first example starts an async Ajax request which will trigger an `alert` in a callback
* The first example finishes before the request finishes (driver resets)
* The callback finishes and triggers the `alert` during the second scenario.

Here's an example debug log:

```
Received "Node.leftClick"
Started "Node.leftClick"
Started request to "http://127.0.0.1:60885/documents/5537d63a6272696b41000002"
Finished "Node.leftClick" with response "Success()"
Wrote response true ""
Received "Reset()"
Started "Reset()"
Finished "Reset()" with response "Success()"
Wrote response true ""
Received 0 from "http://127.0.0.1:60885/documents/5537d63a6272696b41000002"
Received "FindXpath(/html/body/*)"
Started "FindXpath(/html/body/*)"
Finished "FindXpath(/html/body/*)" with response "Success()"
Wrote response true ""
(alert pops up, driver crashes)
```

My best guess is that:

* The `WebPage` instance is deleted, and somehow the `alert` invokes a method on the deleted instance
* For some reason, this results in calling the default `QWebPage` behavior, popping up an alert
* Because the `WebPage` is deleted and being interacted with, we get a null pointer crash

We may need to be more aggressive about making sure that requests are killed during a `reset`. I'm trying to come up with a plausible unit test now.

**Comment by jferris** [comment]
Calling `abort` on any pending `QNetworkReply` instances before deleting the `WebPage` instances seems to do the trick. See #767.

**Comment by jferris** [comment]
I just released 1.5.1, which includes a fix for this issue as I encountered it.

**Comment by Koronen** [comment]
Thanks @jferris! :+1:
